armi/reactor/components/component.py

```diff
@@ -738,26 +757,62 @@ def setNumberDensities(self, numberDensities):
         """
-        self.p.numberDensities = numberDensities
+        self.updateNumberDensities(numberDensities, wipe=True)
 
-    def updateNumberDensities(self, numberDensities):
+    def updateNumberDensities(self, numberDensities, wipe=False):
         """
         Set one or more multiple number densities. Leaves unlisted number densities alone.
 
         Parameters
         ----------
         numberDensities : dict
             nucName: ndens pairs.
+        wipe : bool, optional
+            Controls whether the old number densities are wiped. Any nuclide densities not
+            provided in numberDensities will be effectively set to 0.0.
 
+        Notes
+        -----
+        Sometimes volume/dimensions can change due to the number density change when the material thermal
+        expansion depends on the component's composition (e.g. its plutonium fraction). In this case, changing the
+        density will implicitly change the area/volume. Since it is very difficult to predict the new dimensions
+        ahead of time, and perturbation/depletion calculations are almost exclusively done assuming constant volume,
+        the densities sent are automatically adjusted to conserve mass with the original dimensions. That is, the
+        component's densities are not exactly as passed, but whatever they would need to be to preserve volume
+        integrated number densities (moles) from the pre-perturbed component's volume/dimensions.
+        Note this has no effect if the material thermal expansion has no dependence on component composition.
+        If this is not desired, `self.p.numberDensities` can be set directly.
         """
+        # prepare to change the densities with knowledge that dims could change due to
+        # material thermal expansion dependence on composition
+        if len(self.p.numberDensities) > 0:
+            dLLprev = (
+                self.material.linearExpansionPercent(Tc=self.temperatureInC) / 100.0
+            )
+            materialExpansion = True
+        else:
+            dLLprev = 0.0
+            materialExpansion = False
+
+        # change the densities
+        if wipe:
+            self.p.numberDensities = {}  # clear things not passed
         self.p.numberDensities.update(numberDensities)
-        # since we're updating the object the param points to but not the param itself, we have to inform
+
+        # check if thermal expansion changed
+        dLLnew = self.material.linearExpansionPercent(Tc=self.temperatureInC) / 100.0
+        if dLLprev != dLLnew and materialExpansion:
+            # the thermal expansion changed so the volume change is happening at same time as
+            # density change was requested. Attempt to make mass consistent with old dims
+            # (since the density change was for the old volume and otherwise mass wouldn't be conserved)
+
+            # enable recalculation of volume, otherwise it uses stored.
```

Comment:
```suggestion
            # the thermal expansion changed so the volume change is happening at same time as
            # density change was requested. Attempt to make mass consistent with old dims (since the
            # density change was for the old volume and otherwise mass wouldn't be conserved).

            # enable recalculation of volume, otherwise it uses stored
```